compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -428,12 +428,24 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                     }
                 }
 
-                err.span_suggestion(
-                    call_span,
-                    &format!("try calling `{}` as a method", ident),
-                    format!("self.{}({})", path_str, args_snippet),
-                    Applicability::MachineApplicable,
-                );
+                if let Some(val) = res.filter(|_| source.is_call()) {
```

Comment:
How does this work?
Looks like `call_has_self_arg` above ensures that `source.is_call()` is true.
So the condition is just `if let Some(val) = res`, i.e. "some resolution found, but it doesn't meet restrictions", which seems irrelevant to the issue that you are trying to solve.

Reply:
@petrochenkov, Thanks for your quick review. Yes, you're right. I was thinking that the  `smart_resolve_context_dependent_help` function would do all the checks. Does that mean the `if let Some(val) = res` is irrelevant, because I do need that `val` in order to call `smart_resolve_context_dependent_help`? Also, for the test stderr in `compiler/rustc_resolve/src/late/diagnostics.rs`, the help messages of `try calling bar as a method` are still there. Doesn't that mean that, in some cases, `source.is_call` is false?

Reply:
>Does that mean the `if let Some(val) = res` is irrelevant, because I do need that `val` in order to call `smart_resolve_context_dependent_help`?

It is irrelevant because the "try calling `{}` as a method" heuristic may work regardless of whether there's no resolution or some inappropriate resolution is found.

>Also, for the test stderr in `src/test/ui/self/suggest-self-2.rs`, the help messages of `try calling bar as a method` are still there. Doesn't that mean that, in some cases, `source.is_call` is false?

It only means that `res` is `None` in that case.
`is_call` should always be true at this point due to the `call_has_self_arg` above.